Fix #3490 Add resume lesson activity [Blocked on #3488]

MaskedCarrot opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fixes #3490 

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [ ] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.


## Final result
#### Mobile
<img src="https://user-images.githubusercontent.com/53645584/127312752-52223154-dce6-4639-a323-2fc2d5b9590a.jpeg" width="200" height="355">  <img src="https://user-images.githubusercontent.com/53645584/127316556-36d4f50a-fcb6-4812-b172-9c1a5b3ca0ec.jpeg" width="355" height="200">

#### Tablet

<img src="https://user-images.githubusercontent.com/53645584/127344982-1a60d6d8-c934-41e4-bfe1-d0efc7ed9850.png" width="300" height="420"> <img src="https://user-images.githubusercontent.com/53645584/127344588-df7bccb6-3698-438c-9683-2a783dbb90fa.png" width="420" height="300">

### Espresso test
#### ResumeLessonActivityTest
![Screenshot from 2021-08-20 14-57-26](https://user-images.githubusercontent.com/53645584/130214763-e94572c7-9a73-4a57-a848-a84ef65bf183.png)

#### ResumeLessonFragmentTest
![Screenshot from 2021-08-20 14-56-19](https://user-images.githubusercontent.com/53645584/130214759-536cb9e2-5e85-4567-afd5-b06a7255250d.png)
![Screenshot from 2021-08-20 15-11-46](https://user-images.githubusercontent.com/53645584/130214765-4a789b1a-1dd9-4ed8-9bff-1fcf8f3fb925.png)

Commit:
moved hint handler to domain layer

yashraj-01 committed:
fixed app layer espresso tests

yashraj-01 committed:
fix app layer robolectric tests

yashraj-01 committed:
fixed domain layer tests

yashraj-01 committed:
Added annotations to test exemptions

yashraj-01 committed:
proto lint fix

Commit:
fixed hint handler for training sessions

Commit:
nit and removed test excemptions

Commit:
added hint tests for config change

Commit:
fixed test file exemptions

Commit:
fixed failing test

yashraj-01 committed:
made HintHandler injectable

yashraj-01 committed:
Merge remote-tracking branch 'upstream/move-hint-handler-to-domain' into move-hint-handler-to-domain

yashraj-01 committed:
fixed ktlint error

yashraj-01 committed:
Added tests for hint handler

MaskedCarrot committed:
Merge branch 'develop' into move-hint-handler-to-domain

Commit:
nits

Commit:
Merge remote-tracking branch 'origin/move-hint-handler-to-domain' into move-hint-handler-to-domain

Commit:
fixed failing build

Commit:
fixed failing build

Commit:
fix build

Commit:
fixed imports

MaskedCarrot committed:
Merge branch 'develop' into move-hint-handler-to-domain

Commit:
completed mechansim to resume exploration

Commit:
nits and improved testing

Commit:
updated exploration.proto

Commit:
removed progress controller from kdoc exemptions

Commit:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

Commit:
added kdoc

Commit:
fixed failing build

Commit:
added resume lesson activity

Commit:
nit fix

Commit:
Merge remote-tracking branch 'origin/develop' into move-hint-handler-to-domain

Commit:
fix failing test

Commit:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

Commit:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

Commit:
fixed failing build

Commit:
Revert "fixed failing build"

This reverts commit 6ba0249be01a4a8f91ecdadb702277b3c086edbd.

Commit:
fixed failing build

Commit:
fix failing build

MaskedCarrot commented:
Comment in the previous PR: #3595 

> Thanks @MaskedCarrot. Took a full pass on my codeowner files & related. Besides the comments I've left, I have a few other questions:
> 
> 1. Is the UI implementation finalized? The text spacing & background colors in your screenshots seem incomplete & inconsistent with other screens in the app.


I created the mobile portrait UI with the help of the mock, mobile landscape and tablet mocks are not available in the PRD so i created those by looking at the other screens in the app (primarily topic info fragment because that is similar to this fragment as it has lessonsThumbanilImageVeiw. )

Also, the image in the PR description may look a bit stretched because the screenshot was too large, so I changed the height and width of the image in markdown to make the image smaller.

## Mock
![Screenshot from 2021-07-28 14-02-14](https://user-images.githubusercontent.com/53645584/129113541-fd8ffd77-3413-472d-8290-01f27e6b901c.png)

Note- The overflow menu in the above mock was removed as per the discussions in PRD doc comments.




@BenHenning  PTAL

MaskedCarrot commented:
The failing tests are

testResumeLessonFragment_lessonThumbnailIsDisplayed()
testResumeLessonFragment_onTablet_lessonThumbnailIsDisplayed()
testResumeLessonFragment_onTablet_configChange_lessonThumbnailIsDisplayed()

I tried this in the following two ways:

changing the XML to show the drawable with src.
changing the file LessonThumbnailImageView to force it to show this drawable.
All tests pass for the first case but all of them fail for the second.

Based on the above observation I think this might be happening because the image does not load in time, to verify this I tried adding multiple testCoroutineDispatchers.runCurrent() but the tests still fail.

MaskedCarrot committed:
updated layout files

BenHenning commented on domain/src/main/java/org/oppia/android/domain/topic/TopicController.kt:
```diff
@@ -143,6 +144,30 @@ class TopicController @Inject constructor(
   }
 
+  /**
+   * Retrieves a chapter given a topic ID, story ID and exploration ID.
```
```suggestion
   * Retrieves a chapter given a topic ID, story ID, and exploration ID.
```

Nit on Oxford comma.

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/topic/TopicController.kt:
```diff
@@ -143,6 +144,30 @@ class TopicController @Inject constructor(
+   * @param explorationId the ID corresponding to the exploration which needs to be returned
+   * @return a [DataProvider] for [ChapterSummary]
+   */
+  fun retrieveChapter(
```
Please add tests for this new method.

MaskedCarrot replied:
added tests for this method

BenHenning commented on app/src/main/java/org/oppia/android/app/resumelesson/ResumeLessonFragment.kt:
```diff
@@ -0,0 +1,91 @@
+
+    /**
+     * Creates new instance of [ResumeLessonFragment].
+     * @param internalProfileId is used by the ResumeLessonFragment to retrieve saved checkpoint
```
Nit: a blank doc line should exist between the at-clause block and the summary fragment. Ditto elsewhere in the PR.

MaskedCarrot replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/resumelesson/ResumeLessonFragment.kt:
```diff
@@ -0,0 +1,91 @@
+import javax.inject.Inject
+
+/** Fragment that allows the user to resume a saved exploration. */
+class ResumeLessonFragment :
+  InjectableFragment() {
```
Should be on one line (in general, prefer to keep all statements on one line unless they need to be wrapped.

Ditto elsewhere in the PR (including for ResumeLessonActivity).

MaskedCarrot replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/resumelesson/ResumeLessonActivity.kt:
```diff
@@ -0,0 +1,105 @@
+      "ResumeLessonActivity.backflow_screen"
+
+    /**
+     *  Returns a new [Intent] to route to [ResumeLessonActivity] for a specified exploration.
```
Nit: extra space before ``Returns``

MaskedCarrot replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/resumelesson/ResumeLessonViewModel.kt:
```diff
@@ -0,0 +1,25 @@
+import javax.inject.Inject
+
+/**
+ * The viewModel that providers [ChapterSummary] and [ExplorationCheckpoint] for the exploration
```
```suggestion
 * The viewModel that provides a [ChapterSummary] and [ExplorationCheckpoint] for the exploration
```

MaskedCarrot replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/resumelesson/ResumeLessonViewModel.kt:
```diff
@@ -0,0 +1,25 @@
+  /** The chapter summary for the exploration that may be resumed. */
+  val chapterSummary = ObservableField(ChapterSummary.getDefaultInstance())
+
+  /** The [ExplorationCheckpoint] that may be used to resume exploration. */
```
```suggestion
  /** The [ExplorationCheckpoint] that may be used to resume the exploration. */
```

MaskedCarrot replied:
done

BenHenning reviewed (changes requested):
Thanks @MaskedCarrot. Looked at all of my codeowner files & some of the related ones. Just had a few comments, but otherwise the PR is pretty close to LGTM from my perspective.

BenHenning commented:
> The failing tests are
> 
> testResumeLessonFragment_lessonThumbnailIsDisplayed()
> testResumeLessonFragment_onTablet_lessonThumbnailIsDisplayed()
> testResumeLessonFragment_onTablet_configChange_lessonThumbnailIsDisplayed()
> 
> I tried this in the following two ways:
> 
> changing the XML to show the drawable with src.
> changing the file LessonThumbnailImageView to force it to show this drawable.
> All tests pass for the first case but all of them fail for the second.
> 
> Based on the above observation I think this might be happening because the image does not load in time, to verify this I tried adding multiple testCoroutineDispatchers.runCurrent() but the tests still fail.

@MaskedCarrot

I missed this earlier. As discussed in chat, I think the issue is that LessonThumbnailImageView (used by ResumeLessonFragmentTest) uses ImageLoader internally and, because that uses Glide, the image won't actually load deterministically without setting Glide up a certain way in the test (which is a bit of a pain and not hermetic or fully reliable). I actually suggest switching to TestImageLoaderModule instead of GlideImageLoaderModule and see if that just works (I think it will). If it doesn't, please let me know since that means additional verification needs to be added.

aggarwalpulkit596 commented:
> > The failing tests are
> > testResumeLessonFragment_lessonThumbnailIsDisplayed()
> > testResumeLessonFragment_onTablet_lessonThumbnailIsDisplayed()
> > testResumeLessonFragment_onTablet_configChange_lessonThumbnailIsDisplayed()
> > I tried this in the following two ways:
> > changing the XML to show the drawable with src.
> > changing the file LessonThumbnailImageView to force it to show this drawable.
> > All tests pass for the first case but all of them fail for the second.
> > Based on the above observation I think this might be happening because the image does not load in time, to verify this I tried adding multiple testCoroutineDispatchers.runCurrent() but the tests still fail.
> 
> @MaskedCarrot
> 
> I missed this earlier. As discussed in chat, I think the issue is that LessonThumbnailImageView (used by ResumeLessonFragmentTest) uses ImageLoader internally and, because that uses Glide, the image won't actually load deterministically without setting Glide up a certain way in the test (which is a bit of a pain and not hermetic or fully reliable). I actually suggest switching to TestImageLoaderModule instead of GlideImageLoaderModule and see if that just works (I think it will). If it doesn't, please let me know since that means additional verification needs to be added.

@BenHenning I'm a bit curious here since the image is part of local drawable and even if we loading it with Glide it should be almost immediate and the test shouldn't fail because of that just as it's not failing in the rest of the places where it's being used.

MaskedCarrot committed:
moved timer to domain

MaskedCarrot committed:
fixed build and nits

MaskedCarrot committed:
merge base branch

MaskedCarrot committed:
moved functions to ExplorationProgress.kt

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
nit

MaskedCarrot committed:
nit

MaskedCarrot committed:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
added tests

MaskedCarrot committed:
updated exemptions

MaskedCarrot committed:
added listener back to test exemptions

MaskedCarrot committed:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
nit

MaskedCarrot committed:
nit

MaskedCarrot committed:
updated proto

MaskedCarrot committed:
nit fixes and added more tests

MaskedCarrot committed:
lint fix

MaskedCarrot committed:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

MaskedCarrot committed:
nit fix

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
nit fix

MaskedCarrot committed:
removed unused fun

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
updated xml

rt4914 commented on app/src/main/res/drawable/ic_start_over_24dp.xml:
```diff
@@ -0,0 +1,9 @@
+  <path
+    android:fillColor="#FFFFFF"
+    android:pathData="M6,6h2v12L6,18zM9.5,12l8.5,6L18,6z"/>
+</vector>
```
EOF is missing, add one blank line at the end.

MaskedCarrot replied:
done

rt4914 commented on app/src/main/res/drawable/start_over_button_background.xml:
```diff
@@ -0,0 +1,8 @@
+  <stroke
+    android:width="4px"
+    android:color="@color/oppiaPrimaryDark" />
+</shape>
```
EOF is missing, add one blank line at the end.

MaskedCarrot replied:
done

rt4914 commented on app/src/main/res/drawable/start_over_button_background.xml:
```diff
@@ -0,0 +1,8 @@
+  android:shape="rectangle">
+  <corners android:radius="@dimen/state_previous_next_button_radius" />
+  <stroke
+    android:width="4px"
```
Why is this `px`? Normally we keep all values as `dp`

MaskedCarrot replied:
done

rt4914 commented on app/src/main/java/org/oppia/android/app/resumelesson/ResumeLessonActivity.kt:
```diff
@@ -0,0 +1,103 @@
+  // TODO(#1655): Re-restrict access to fields in tests post-Gradle.
+  companion object {
+    private const val RESUME_LESSON_ACTIVITY_PROFILE_ID_ARGUMENT_KEY =
+      "ResumeLessonActivity.profile_id"
```
Instead of `profileId`/ `ProfileId`/`PROFILE_ID`, etc. use `internalProfileId`/`InternalProfileId`/`INTERNAL_PROFILE_ID` everywhere.

Basically when its `Int` we use `internalProfileId` as `ProfileId` is actually a model in our proto.

MaskedCarrot replied:
done

rt4914 reviewed (changes requested):
@MaskedCarrot PLAT at following points.
1. Refer to [mock](https://xd.adobe.com/view/3dca36c2-5115-419c-b25e-0f10526b077c-6899/screen/1c7d4ec6-a5f1-4462-886d-e47c1457b69e/) for current xml values.
2. File two issues (1.) for Highfi-Mobile-Landscape and (2.) Highfi Tablet and mark both of them as blocked and then send an email to Chantel asking for mock. Keep me, Ben, Sean, Pulkit in cc.
3. Please add `RTL` screenshots too for all LTR images so that we can verify that it works correctly on RTL too.

BenHenning committed:
First round: make HintHandler independent.

This also brings HintHandler into an interface + factory pattern. This
isn't the final design since I think we can largely simplify the way
hints work; that'll be my next pass.

This breaks questions & HintHandlerTest; those will require further work
later.

BenHenning committed:
Simplify HelpIndex in proto.

Move HelpIndex to PendingState to avoid the entire domain case of
handling CompletedState.

BenHenning committed:
Simplify hints & solutions.

This removes the per-Hint/Solution tracking & completely leans on
HelpIndex for proper hints & solution tracking both in the domain & UI
layers.

Fixes a bunch of other stuff, too, including
QuestionAssessmentProgressController tests.

BenHenning committed:
Clean up dead code paths & improve handler API.

MaskedCarrot commented:
@rt4914 using the exact values from the mock, the button doesn't show the full text. Can you please take a look
![Screenshot from 2021-08-18 00-09-14](https://user-images.githubusercontent.com/53645584/129782752-43d5e63a-e1f7-40a7-881f-89791ee92496.png)

MaskedCarrot committed:
updated xml and nits

yashraj-01 committed:
fixed test modules and lint

yashraj-01 committed:
renamed HintHandlerTest to HintHandlerImplTest

rt4914 commented:
> @rt4914 using the exact values from the mock, the button doesn't show the full text. Can you please take a look
> ![Screenshot from 2021-08-18 00-09-14](https://user-images.githubusercontent.com/53645584/129782752-43d5e63a-e1f7-40a7-881f-89791ee92496.png)

@MaskedCarrot Actually you will need to write attributes in a different way.
layout_width = wrap_content
layout_height = wrap_content
paddingStart = 12dp
paddingEnd = 12dp
For above two paddings you might need to add (drawablePadding) too as you have a drawable in this button.
Also make `min_height` = "48dp". This won't match correctly with the mocks but we need this to make sure buttons are correctly accessible.

BenHenning committed:
Add tests for HintHandler.

This introduces some new explorations for making testing HintHandlerImpl
easier.

MaskedCarrot committed:
updated xml

BenHenning committed:
Add remaining tests/exemptions for new files.

BenHenning committed:
Lint fixes.

MaskedCarrot commented:
@rt4914 PTAL, I have used a fixed layout_height of 144dp instead of wrap content, because since the text lengths in the two buttons are different using wrap content will make them have unequal lengths

MaskedCarrot committed:
nit

BenHenning committed:
Merge branch 'develop' into move-hint-handler-to-domain

Conflicts:
	app/src/main/java/org/oppia/android/app/application/ApplicationComponent.kt
	app/src/sharedTest/java/org/oppia/android/app/home/TopicSummaryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/WelcomeViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryListViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/parser/CustomBulletSpanTest.kt
	app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
	app/src/sharedTest/java/org/oppia/android/app/splash/SplashActivityTest.kt
	app/src/test/java/org/oppia/android/app/home/HomeActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/player/exploration/ExplorationActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/profile/ProfileChooserFragmentLocalTest.kt
	app/src/test/java/org/oppia/android/app/testing/CompletedStoryListSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/HomeSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/OngoingTopicListSpanTest.kt
app/src/test/java/org/oppia/android/app/testing/ProfileChooserSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/ProfileProgressSpanCount.kt
	app/src/test/java/org/oppia/android/app/testing/RecentlyPlayedSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/TopicRevisionSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/player/split/PlayerSplitScreenTesting.kt
	app/src/test/java/org/oppia/android/app/testing/player/state/StateFragmentAccessibilityTest.kt
	app/src/test/java/org/oppia/android/app/topic/revisioncard/RevisionCardActivityLocalTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionTrainingControllerTest.kt

BenHenning committed:
Post-merge fixes (including lint fixes).

BenHenning committed:
Merge branch 'develop' into move-hint-handler-to-domain

Conflicts:
	third_party/maven_install.json

BenHenning committed:
Post-merge maven_install fix.

MaskedCarrot committed:
Merge branch 'develop' into move-hint-handler-to-domain

BenHenning commented on app/src/main/java/org/oppia/android/app/resumelesson/ResumeLessonFragmentPresenter.kt:
```diff
@@ -133,6 +133,7 @@ class ResumeLessonFragmentPresenter @Inject constructor(
       fragment,
       Observer<ChapterSummary> { chapterSummary ->
         resumeLessonViewModel.chapterSummary.set(chapterSummary)
+        val t = chapterSummary
```
Unused val?

MaskedCarrot replied:
removed

BenHenning commented on domain/src/main/java/org/oppia/android/domain/topic/TopicController.kt:
```diff
@@ -156,16 +161,18 @@ class TopicController @Inject constructor(
-          topicId,
-          storyId,
-          explorationId
+  ): LiveData<AsyncResult<ChapterSummary>> {
```
Please don't do this--domain controllers should return DataProviders. LiveData are for the UI layer, not the domain layer.

MaskedCarrot replied:
updated retrieveChapter

BenHenning commented on domain/src/main/java/org/oppia/android/domain/topic/TopicController.kt:
```diff
@@ -156,16 +161,18 @@ class TopicController @Inject constructor(
+      try {
+        AsyncResult.success(fetchChapter(topicId, storyId, explorationId))
+      } catch (e: Exception) {
+        AsyncResult.failed(
+          ChapterNotFoundException(
+            "Chapter for exploration $explorationId not found in story $storyId and topic $topicId"
+          )
```
If you need a failure case then I suggest returning AsyncResult.failure() in the original DataProvider.

MaskedCarrot replied:
updated retrieveChapter

BenHenning commented on domain/src/main/java/org/oppia/android/domain/topic/TopicController.kt:
```diff
@@ -95,6 +95,11 @@ class TopicController @Inject constructor(
 ) {
 
+  /**
+   * Indicates that chapter for the specified exploration Id, story ID, and topicID was not found.
```
```suggestion
   * Indicates that the chapter for the specified exploration, story, and topic ID was not found.
```

Nits on keeping consistency when referring to structure IDs.

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/topic/TopicController.kt:
```diff
@@ -156,16 +161,18 @@ class TopicController @Inject constructor(
+    return MutableLiveData(
+      try {
+        AsyncResult.success(fetchChapter(topicId, storyId, explorationId))
+      } catch (e: Exception) {
```
We shouldn't be catching a generic exception like this--how do you know the correct one was caught? Also, it's considered an antipattern to use exceptions for control flow. Instead, return a result or introduce a mechanism to check for existence to make the determination that the chapter doesn't exist.

MaskedCarrot replied:
updated retrieveChapter

MaskedCarrot replied:
updated retrieveChapter

BenHenning reviewed (changes requested):
Thanks @MaskedCarrot. Took another pass. PTAL.

I'll need to review the TopicControllerTest changes more closely after this latest round of comments is addressed.

> > > The failing tests are
> > > testResumeLessonFragment_lessonThumbnailIsDisplayed()
> > > testResumeLessonFragment_onTablet_lessonThumbnailIsDisplayed()
> > > testResumeLessonFragment_onTablet_configChange_lessonThumbnailIsDisplayed()
> > > I tried this in the following two ways:
> > > changing the XML to show the drawable with src.
> > > changing the file LessonThumbnailImageView to force it to show this drawable.
> > > All tests pass for the first case but all of them fail for the second.
> > > Based on the above observation I think this might be happening because the image does not load in time, to verify this I tried adding multiple testCoroutineDispatchers.runCurrent() but the tests still fail.
> > 
> > 
> > @MaskedCarrot
> > I missed this earlier. As discussed in chat, I think the issue is that LessonThumbnailImageView (used by ResumeLessonFragmentTest) uses ImageLoader internally and, because that uses Glide, the image won't actually load deterministically without setting Glide up a certain way in the test (which is a bit of a pain and not hermetic or fully reliable). I actually suggest switching to TestImageLoaderModule instead of GlideImageLoaderModule and see if that just works (I think it will). If it doesn't, please let me know since that means additional verification needs to be added.
> 
> @BenHenning I'm a bit curious here since the image is part of local drawable and even if we loading it with Glide it should be almost immediate and the test shouldn't fail because of that just as it's not failing in the rest of the places where it's being used.

@aggarwalpulkit596 

I think you might have a false assumption here (that loading a local drawable via Glide is immediate--it's not). Glide's pipeline has a lot of black boxing to it including caching & potentially utilizing custom loading pipelines, even for local files (which we see with SVGs, for example). It seems Glide makes heavy use of its executors for all image loading, and getting these to behave deterministically in the test environment is tricky.

MaskedCarrot committed:
Revert "Merge branch 'develop' into move-hint-handler-to-domain"

This reverts commit e2fea908a65af8c11b5ed135f59cec38e807f251, reversing
changes made to 6659858f0b85a0bf0c51f23af600769eac64b938.

BenHenning committed:
Post-merge Gradle-discovered fixes.

BenHenning committed:
Revert "Revert "Merge branch 'develop' into move-hint-handler-to-domain""

This reverts commit b1622c00d33f584ae727f0c9c4434ad773bc80a2.

BenHenning committed:
Merge branch 'attempt-to-fix-move-hint-handler-to-domain' into move-hint-handler-to-domain

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/home/TopicSummaryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/WelcomeViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryListViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/parser/CustomBulletSpanTest.kt
	app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
	app/src/sharedTest/java/org/oppia/android/app/splash/SplashActivityTest.kt
	app/src/test/java/org/oppia/android/app/home/HomeActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/player/exploration/ExplorationActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/profile/ProfileChooserFragmentLocalTest.kt
	app/src/test/java/org/oppia/android/app/testing/CompletedStoryListSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/HomeSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/OngoingTopicListSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/ProfileChooserSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/ProfileProgressSpanCount.kt
	app/src/test/java/org/oppia/android/app/testing/RecentlyPlayedSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/TopicRevisionSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/player/split/PlayerSplitScreenTesting.kt
	app/src/test/java/org/oppia/android/app/testing/player/state/StateFragmentAccessibilityTest.kt
	app/src/test/java/org/oppia/android/app/topic/revisioncard/RevisionCardActivityLocalTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionTrainingControllerTest.kt

BenHenning committed:
Additional post-merge fixes.

MaskedCarrot committed:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

MaskedCarrot committed:
updated appraoch to resume hints to work with the new hint handler.

MaskedCarrot committed:
nit fix

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
updated fragment component

MaskedCarrot committed:
updated topic controller

rt4914 commented on app/src/main/res/values/styles.xml:
```diff
@@ -327,4 +327,39 @@
+    <item name="android:layout_height">wrap_content</item>
+    <item name="android:drawablePadding">4dp</item>
+    <item name="android:paddingRight">12dp</item>
+    <item name="android:paddingLeft">12dp</item>
```
Please use `paddingStart` and `paddingEnd` here and everywhere else in this PR.

Left/Right does not translate well for RTL

MaskedCarrot replied:
done

rt4914 commented on app/src/main/res/values/styles.xml:
```diff
@@ -327,4 +327,39 @@
+    <item name="android:layout_width">144dp</item>
+    <item name="android:layout_height">wrap_content</item>
+    <item name="android:paddingRight">12dp</item>
+    <item name="android:paddingLeft">12dp</item>
```
Do not mention paddingTop and paddingBottom. Instead mention `minHeight = 48dp` as we need this to make this correctly accessible

MaskedCarrot replied:
done

rt4914 commented on app/src/main/res/values/styles.xml:
```diff
@@ -327,4 +327,39 @@
+    <item name="android:paddingRight">12dp</item>
+    <item name="android:paddingLeft">12dp</item>
+    <item name="android:paddingTop">10dp</item>
+    <item name="android:paddingBottom">10dp</item>
```
Do not mention paddingTop and paddingBottom. Instead mention `minHeight = 48dp` as we need this to make this correctly accessible

MaskedCarrot replied:
done

rt4914 commented on app/src/main/res/values/styles.xml:
```diff
@@ -327,4 +327,39 @@
+  <style name="StartOverLessonButton" parent="TextAppearance.AppCompat.Widget.Button">
+    <item name="android:layout_width">144dp</item>
+    <item name="android:layout_height">wrap_content</item>
+    <item name="android:paddingRight">12dp</item>
```
Please use paddingStart and paddingEnd here and everywhere else in this PR.

Left/Right does not translate well for RTL

MaskedCarrot replied:
done

rt4914 reviewed (changes requested):
@MaskedCarrot PTAL

MaskedCarrot committed:
updated xml and tests

MaskedCarrot commented:
@BenHenning Can you please also take a look at the test `fun testRetrieveChapter_validChapter_returnsCorrectChapterSummary()` present in `topicControllerTest`. 

This test is currently failing on Bazel but passing on Gradle.

This test checks the following: 
1. chapter name 
2. chapter summary 
3. chapter thumbnail
out of the 3 the test only fails for the last one (chapter thumbnail) with the error message expected was 4 actual was 0.


[link to the test](https://github.com/oppia/oppia-android/blob/4b68d28882a7c8c5e778f38770487ecb9db943a2/domain/src/test/java/org/oppia/android/domain/topic/TopicControllerTest.kt#L416)

MaskedCarrot committed:
nit

MaskedCarrot committed:
nit

MaskedCarrot committed:
nit

MaskedCarrot committed:
nit

BenHenning commented on domain/src/main/java/org/oppia/android/domain/topic/TopicController.kt:
```diff
@@ -161,18 +161,20 @@ class TopicController @Inject constructor(
+      return@createInMemoryDataProviderAsync AsyncResult.success(retrieveStory(topicId, storyId))
+    }.transformAsync(GET_CHAPTER_PROVIDER_ID) { storySummary ->
+      val chapterSummary = fetchChapter(storySummary, explorationId)
+      if (chapterSummary == null)
```
This if/else needs braces per https://developer.android.com/kotlin/style-guide#braces.

BenHenning replied:
A nit: but I also suggest inverting the if/else condition so that the default cause is success (since that's the expected happy path).

MaskedCarrot replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/topic/TopicControllerTest.kt:
```diff
@@ -431,19 +430,11 @@ class TopicControllerTest {
-      .isEqualTo(
-        "Chapter for exploration $INVALID_EXPLORATION_ID_1 not found in story " +
-          "$INVALID_STORY_ID_1 and topic $INVALID_TOPIC_ID_1"
-      )
```
It's probably still worth verifying that AsyncResult includes the correct failure.

MaskedCarrot replied:
updated the test to verify the error on failure.

BenHenning reviewed (changes requested):
Thanks @MaskedCarrot! Just had a few comments, otherwise the PR LGTM (barring any issues when upstream PRs are merged).

> @BenHenning Can you please also take a look at the test `fun testRetrieveChapter_validChapter_returnsCorrectChapterSummary()` present in `topicControllerTest`.
> 
> This test is currently failing on Bazel but passing on Gradle.
> 
> This test checks the following:
> 
> 1. chapter name
> 2. chapter summary
> 3. chapter thumbnail
>    out of the 3 the test only fails for the last one (chapter thumbnail) with the error message expected was 4 actual was 0.
> 
> [link to the test](https://github.com/oppia/oppia-android/blob/4b68d28882a7c8c5e778f38770487ecb9db943a2/domain/src/test/java/org/oppia/android/domain/topic/TopicControllerTest.kt#L416)

This is probably because TopicControllerTest uses the textproto versions of the test explorations in tests when run on Bazel & the JSON versions when run in Gradle. The Bazel versions intentionally dropped the temporary thumbnails, and in cases when we test with both types we usually don't validate thumbnails for this reason.

Is verifying the name sufficient for your test? I suggest dropping the thumbnail assertion altogether if it the test is still valid without it. If you need to verify the thumbnail, reach out to me on chat and we'll try to figure out a solution.

MaskedCarrot committed:
Merge remote-tracking branch 'origin/develop' into resume-exploration-in-domain-module

MaskedCarrot committed:
fixed merge

MaskedCarrot committed:
test and lint fix

rt4914 reviewed (approved):
LGTM, thanks

MaskedCarrot committed:
updated

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
nit fix

MaskedCarrot committed:
lint fix

MaskedCarrot committed:
fix failing tests

MaskedCarrot committed:
fixed tests

MaskedCarrot committed:
lint fix

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

vinitamurthi commented:
I dont think I am required here because I dont own any of these files, removing myself as a reviewer!

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/resumelesson/ResumeLessonActivityTest.kt:
```diff
@@ -0,0 +1,233 @@
+  application = ResumeLessonActivityTest.TestApplication::class,
+  qualifiers = "port-xxhdpi"
+)
+class ResumeLessonActivityTest {
```
add kdoc and espresso test result screenshot

MaskedCarrot replied:
done

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/resumelesson/ResumeLessonActivityTest.kt:
```diff
@@ -0,0 +1,233 @@
+          instanceOf(TextView::class.java),
+          withParent(withId(R.id.resume_lesson_activity_toolbar))
+        )
+      )
+        .check(matches(withText("Resume Lesson")))
```
```suggestion
      ).check(matches(withText("Resume Lesson")))
```

MaskedCarrot replied:
done

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/resumelesson/ResumeLessonActivityTest.kt:
```diff
@@ -0,0 +1,233 @@
+          instanceOf(TextView::class.java),
+          withParent(withId(R.id.resume_lesson_activity_toolbar))
+        )
+      )
+        .check(matches(withText("Resume Lesson")))
```
ditto

MaskedCarrot replied:
done

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/resumelesson/ResumeLessonFragmentTest.kt:
```diff
@@ -0,0 +1,260 @@
+  application = ResumeLessonFragmentTest.TestApplication::class,
+  qualifiers = "port-xxhdpi"
+)
+class ResumeLessonFragmentTest {
```
add kdoc and espresso result screenshot for mobile and tablet

MaskedCarrot replied:
done

anandwana001 reviewed (changes requested):
left few points.

MaskedCarrot committed:
nit

MaskedCarrot committed:
lint fix

MaskedCarrot committed:
Merge branch 'develop' into resume-exploration-in-domain-module

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
improved testing and nit fix

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/resumelesson/ResumeLessonFragmentTest.kt:
```diff
@@ -119,7 +121,7 @@ class ResumeLessonFragmentTest {
     ApplicationProvider.getApplicationContext<TestApplication>().inject(this)
   }
-
+  @Ignore
```
Newline before this (newlines generally always separate class members except fields which are optional).

MaskedCarrot replied:
ignore is not required, this test is for tablets and it fails on mobile, so I added this while testing the tests on espresso

MaskedCarrot replied:
removed ignore

BenHenning reviewed (commented):
Thanks @MaskedCarrot. Just 1 nit. PTAL.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/resumelesson/ResumeLessonFragmentTest.kt:
```diff
@@ -0,0 +1,262 @@
+  private fun setUpTestApplicationComponent() {
+    ApplicationProvider.getApplicationContext<TestApplication>().inject(this)
+  }
+  @Ignore
```
Why are we ignoring this test case?
Add reason in format `@Ignore TODO ... ` and add TODO of issue which will fix this.

MaskedCarrot replied:
ignore is not required, this test is for tablets and it fails on mobile, so I added this while testing the tests on espresso

MaskedCarrot replied:
removed this

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
nit

MaskedCarrot committed:
Merge remote-tracking branch 'origin/develop' into resume-exploration-in-domain-module

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
nit

MaskedCarrot committed:
nit

MaskedCarrot committed:
nit

MaskedCarrot committed:
fix failing test

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
fix tests

MaskedCarrot committed:
fixed tests

BenHenning reviewed (commented):
Thanks @MaskedCarrot. LGTM, but waiting on approving until it's based on develop with passing CI. Please assign back at that point.

MaskedCarrot committed:
Merge remote-tracking branch 'origin/develop' into add-resume-lesson-activity

anandwana001 reviewed (approved):
Approving my last suggestion got fixed. LGTM

BenHenning reviewed (approved):
Thanks @MaskedCarrot. LGTM!

BenHenning commented:
Merging since everyone has approved, CI is passing, branch is up-to-date & looks good, and no conversation threads are open anymore.
